toolkit/primitives/block-participation/src/lib.rs

```diff
@@ -0,0 +1,370 @@
+#![cfg_attr(not(feature = "std"), no_std)]
+
+extern crate alloc;
+
+use alloc::vec::Vec;
+use parity_scale_codec::{Decode, Encode};
+use scale_info::TypeInfo;
+use sidechain_domain::{MainchainKeyHash, McEpochNumber};
+pub use sp_consensus_slots::{Slot, SlotDuration};
+use sp_inherents::{InherentIdentifier, IsFatalError};
+
+#[cfg(test)]
+mod tests;
+
+pub const INHERENT_IDENTIFIER: InherentIdentifier = *b"blokpart";
+
+#[derive(Clone, Debug, PartialEq, Eq, Decode, Encode, TypeInfo, PartialOrd, Ord)]
+pub struct DelegatorBlockParticipationData<DelegatorId> {
+	id: DelegatorId,
+	share: u64,
+}
+
+#[derive(Clone, Debug, PartialEq, Eq, Decode, Encode, TypeInfo, PartialOrd, Ord)]
+pub struct BlockProducerParticipationData<BlockProducerId, DelegatorId> {
```

Comment:
Maybe I'm not on form, but I struggle to understand the code and additional type parameters don't help me at all. We provide Cardano data, where `BlockProducerId` is a **Stake Pool Key Hash** and `DelegatorId` is **Staking Key Hash**.